File: src/java/tree/110.balanced-binary-tree.java
class P110 {
    // @lc app=leetcode id=110 lang=java
    // @lc code=start
    public class Solution {
        public boolean isBalanced(TreeNode root) { return height(root) != -1; }
        private int height(TreeNode n){
            if (n==null) return 0;
            int lh = height(n.left); if (lh==-1) return -1;
            int rh = height(n.right); if (rh==-1) return -1;
            if (Math.abs(lh-rh) > 1) return -1;
            return 1 + Math.max(lh, rh);
        }
    }
    // @lc code=end
}

File: src/python/tree/110.balanced-binary-tree.py
from typing import Optional
from lc_test_utils import TreeNode

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def check(node):
            if not node:
                return 0
            
            l = check(node.left)
            r = check(node.right)
            
            if l == -1 or r == -1 or abs(l - r) > 1:
                return -1
            return 1 + max(l, r)
            
        return check(root) != -1
